app/main/views/suppliers.py

```diff
@@ -165,75 +165,65 @@ def edit_registered_address():
     except APIError as e:
         abort(e.status_code)
 
-    if supplier_form is None:
-        supplier_form = EditSupplierForm(
-            description=supplier.get('description', None),
-            clients=supplier.get('clients', None)
-        )
-        contact_form = EditContactInformationForm(
-            prefix='contact_',
-            **supplier['contactInformation'][0]
-        )
-
-    return render_template(
-        "suppliers/edit_supplier.html",
-        error=error,
-        supplier_form=supplier_form,
-        contact_form=contact_form
-    ), 200
+    supplier['contact'] = supplier['contactInformation'][0]
+    error = None
+    http_status = 200
 
+    supplier_form = EditSupplierForm()
+    contact_form = EditContactInformationForm()
 
-@main.route('/details/edit', methods=['POST'])
-@login_required
-def update_supplier():
-    # FieldList expects post parameter keys to have number suffixes
-    # (eg client-0, client-1 ...), which is incompatible with how
-    # JS list-entry plugin generates input names. So instead of letting
-    # the form search for request keys we pass in the values directly as data
-    supplier_form = EditSupplierForm(
-        formdata=None,
-        description=request.form['description'],
-    )
+    if request.method == 'POST':
+        supplier_info_valid = supplier_form.validate_on_submit()
+        contact_info_valid = contact_form.validate_on_submit()
 
-    contact_form = EditContactInformationForm(prefix='contact_')
+        if supplier_info_valid and contact_info_valid:
+            try:
+                data_api_client.update_supplier(
+                    current_user.supplier_id,
+                    supplier_form.data,
+                    current_user.email_address
+                )
 
-    if not (supplier_form.validate_on_submit() and contact_form.validate_on_submit()):
-        return edit_supplier(supplier_form=supplier_form, contact_form=contact_form)
+                data_api_client.update_contact_information(
+                    current_user.supplier_id,
+                    supplier['contact']['id'],
+                    contact_form.data,
+                    current_user.email_address
+                )
+            except APIError as e:
+                error = e.message
+            else:
+                return redirect(url_for(".supplier_details"))
 
-    try:
-        data_api_client.update_supplier(
-            current_user.supplier_id,
-            supplier_form.data,
-            current_user.email_address
-        )
+        http_status = 400
 
-        data_api_client.update_contact_information(
-            current_user.supplier_id,
-            contact_form.id.data,
-            contact_form.data,
-            current_user.email_address
-        )
-    except APIError as e:
-        return edit_supplier(supplier_form=supplier_form,
-                             contact_form=contact_form,
-                             error=e.message)
+    else:
```

Comment:
I don't think you need the else here?

Reply:
I think we might, to avoid overwriting the data if the forms do not validate on POST request.